pkg/trustedresources/verify.go

```diff
@@ -100,18 +91,81 @@ func VerifyPipeline(ctx context.Context, pipelineObj v1beta1.PipelineObject, k8s
 		ObjectMeta: pm,
 		Spec:       pipelineObj.PipelineSpec(),
 	}
-	verifiers, err := getVerifiers(ctx, k8s)
+
+	return verifyResource(ctx, &pipeline, k8s, signature, source, verificationPolicies)
+}
+
+// verifyResource verifies a resource
+func verifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.Interface, signature []byte, source string, verificationPolicies []*v1alpha1.VerificationPolicy) error {
+	if len(verificationPolicies) != 0 {
+		// get all matching policies
+		matchedPolicies := []*v1alpha1.VerificationPolicy{}
+		for _, vp := range verificationPolicies {
+			match := false
+			for _, ra := range vp.Spec.Resources {
+				matching, _ := regexp.MatchString(ra.Pattern, source)
+				// If not matching then skip fetching keys
+				if matching {
+					match = true
+					break
+				}
+			}
+			if match {
+				matchedPolicies = append(matchedPolicies, vp)
+			}
+		}
+		// verify the resource on matchedPolicies
+		verified := 0
+		for _, mp := range matchedPolicies {
+			verifiers, err := getVerifiersFromPolicy(ctx, k8s, source, mp)
+			if err != nil {
+				return err
+			}
+			for _, verifier := range verifiers {
+				// if one of the verifier passes verification, then this policy passes verification
+				if err := verifyInterface(resource, verifier, signature); err == nil {
+					verified++
+					break
+				}
+			}
+		}
+		if len(matchedPolicies) == 0 {
+			return fmt.Errorf("No matching policies found for resource %s from source %s", resource.GetName(), source)
+		}
+		// make sure all matched policies pass verification
+		if verified == len(matchedPolicies) {
```

Comment:
I think this check assumes that verifiers are 1:1 with matchedPolicies. Could we have situations where matchedPolicy[0] has 2 successful verifiers but matchPolicy[1] has none?

(not sure what the expected behavior is - but either way I don't think this logic is quite representing what we want)

Reply:
` Could we have situations where matchedPolicy[0] has 2 successful verifiers but matchPolicy[1] has none?`
This is not allowed, so when looping over the keys, if one key passes we will break the loop. I added a test where 1 policy has 1 working key, the other policy has 2 working keys to make sure the behaviour is what we want

Reply:
verifiers shouldn't be not 1:1 with matchedPolicies. 
The current logic is:
```
for p in matchedpolicy{
     for key in p{
         if key works{
         verified++
          break
           }
    }
}
```